packages/netlify-cms-core/src/reducers/entryHistory.js

```diff
@@ -0,0 +1,63 @@
+import { Map } from 'immutable';
+import { ADD_ENTRY_HISTORY } from 'Actions/entryHistory';
+import { ENTRY_PERSIST_SUCCESS, ENTRY_DELETE_SUCCESS } from 'Actions/entries';
+
+const defaultHandler = state => state;
+
+const newHistoryForEntry = (state, { payload }) => {
+  const { key, commits } = payload;
+  return state.set(key, commits);
+};
+
+const entryPersistSuccess = (state, { payload }) => {
+  const { collectionName, slug, newRef, author, commitDate, oldRef, revert } = payload;
+  const key = `${collectionName}.${slug}`;
+
+  const newHistoryObject = {
+    ref: newRef,
+    date: commitDate,
+    author,
+    file: {
+      // TODO REVISIT THIS - THIS MIGHT BE VERY WRONG. (Even if it isn't really being used I think -> in that case remove);
+      path: `_${collectionName}/${slug}.md`
+    }
+  };
+
+  return state.withMutations(mutatingState => {
+    mutatingState.updateIn([key], history => {
+      const newHistory = history ? history.concat([newHistoryObject]) : [newHistoryObject];
+      return revert ? newHistory.filter(({ ref }) => ref !== oldRef) : newHistory;
+    });
+  });
+};
+
+const entryDeleteSuccess = (state, { payload }) => {
```

Comment:
could be a one-liner if you destructure payload in the arguments and use an implicit return :)